beacon-chain/p2p/gossipsub_peer_crawler.go

```diff
@@ -0,0 +1,550 @@
+		cp.peerNodeByPid[peerID] = existingPNode
+	} else {
+		existingPNode.node = node
+	}
+
+	cp.updateTopicsUnlocked(existingPNode, topics)
+
+	if existingPNode.isPinged || len(topics) == 0 {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (cp *crawledPeers) removeTopic(topic string) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	// Get all peers subscribed to this topic
+	peers, ok := cp.pidsByTopic[topic]
+	if !ok {
+		return // Topic doesn't exist
+	}
+
+	// Remove the topic from each peer's topic list
+	for peerID := range peers {
+		if pnode, ok := cp.peerNodeByPid[peerID]; ok {
+			delete(pnode.topics, topic)
+			// remove the peer if it has no more topics left
+			if len(pnode.topics) == 0 {
+				cp.updateTopicsUnlocked(pnode, nil)
+			}
+		}
+	}
+
+	// Remove the topic from byTopic map
+	delete(cp.pidsByTopic, topic)
+}
+
+func (cp *crawledPeers) removePeerByPeerId(peerID peer.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	pnode, ok := cp.peerNodeByPid[peerID]
+	if !ok {
+		return
+	}
+
+	// Use updateTopicsUnlocked with empty topics to remove the peer
+	cp.updateTopicsUnlocked(pnode, nil)
+}
+
+func (cp *crawledPeers) removePeerByNodeId(enodeID enode.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+	pnode, ok := cp.peerNodeByEnode[enodeID]
+	if !ok {
+		return
+	}
+	cp.updateTopicsUnlocked(pnode, nil)
+}
+
+func (cp *crawledPeers) cleanupPeer(pnode *peerNode) {
+	delete(cp.peerNodeByPid, pnode.peerID)
+	delete(cp.peerNodeByEnode, pnode.node.ID())
+	for t := range pnode.topics {
+		if peers, ok := cp.pidsByTopic[t]; ok {
+			delete(peers, pnode.peerID)
+			if len(peers) == 0 {
+				delete(cp.pidsByTopic, t)
+			}
+		}
+	}
+	pnode.topics = nil // Clear topics to indicate removal.
+}
+
+func (cp *crawledPeers) removeOldTopicsFromPeer(pnode *peerNode, newTopics map[string]struct{}) {
+	for oldTopic := range pnode.topics {
+		if _, ok := newTopics[oldTopic]; !ok {
+			if peers, ok := cp.pidsByTopic[oldTopic]; ok {
+				delete(peers, pnode.peerID)
+				if len(peers) == 0 {
+					delete(cp.pidsByTopic, oldTopic)
+				}
+			}
+		}
+	}
+}
+
+func (cp *crawledPeers) addNewTopicsToPeer(pnode *peerNode, newTopics map[string]struct{}) {
+	for newTopic := range newTopics {
+		if _, ok := pnode.topics[newTopic]; !ok {
+			if _, ok := cp.pidsByTopic[newTopic]; !ok {
+				cp.pidsByTopic[newTopic] = make(map[peer.ID]struct{})
+			}
+			cp.pidsByTopic[newTopic][pnode.peerID] = struct{}{}
+		}
+	}
+}
+
+// updateTopicsUnlocked updates the topics associated with a peer node.
+// If the topics slice is empty, the peer is completely removed from the crawled peers.
+// Otherwise, it updates the peer's topics by removing old topics that are no longer
+// present and adding new topics. This method assumes the caller holds the lock on cp.mu.
+// If a topic has no peers after this update, it is removed from the list of topics we track peers for.
+func (cp *crawledPeers) updateTopicsUnlocked(pnode *peerNode, topics []string) {
```

Comment:
I'm not opposed to have both `peerNode.topics` and `cp.pidsByTopic[<topic>]`.
However, I would like to be sure these two items cannot contradict each other, in any way.
Currently, it is possible to have a contradiction.

Reply:
I've changed this to `cp.peersByTopic[topic]*peerNode` as we discussed above.